contracts/Unipool.sol

```diff
@@ -135,8 +149,101 @@ contract Unipool is LPTokenWrapper {
+
+        uint256 minimumTokens = _getMinimumTokensForRewardSale(rewardToConvert);
+        uint[] memory amounts = uniswapRouter.swapExactTokensForTokens(
+            rewardToConvert, 
+            minimumTokens, 
+            path, 
+            address(this), 
+            block.timestamp);
+
+        // We exchanged this amount of the reward token in the swap, remove it from the current balance
+        rewardAmount = rewardAmount.sub(amounts[0]);
+
+        // Use our whole balance of `tradableToken`, not just what we got back from the swap.
+        // This makes sure that shavings of `tradableToken` don't accumulate, dead, in our wallet.
+        // The next user to call this method will always either leave a few shavings themselves,
+        // or will receive the balance back in `rewardToken` instead, which we do account for per-user.
+        uint256 reinvestableBalance = tradableToken.balanceOf(address(this));
+
+        // Approve the amount we're going to try and add to the pool
+        rewardToken.approve(address(uniswapRouter), rewardToConvert);
+        tradableToken.approve(address(uniswapRouter), reinvestableBalance);
+        
+        // add liquidity
+        (uint rewardTokenAdded, , uint liquidity) = 
+            uniswapRouter.addLiquidity(address(rewardToken),
+                                       address(tradableToken),
+                                       rewardToConvert,
+                                       reinvestableBalance,
+                                       0, 0, // min
+                                       address(this),
+                                       block.timestamp);
+        
+        // revoke approval
+        rewardToken.approve(address(uniswapRouter), 0);
+        tradableToken.approve(address(uniswapRouter), 0);
+
+        // now also remove the amount of liquidity we added for the final total
+        return (liquidity, rewardAmount.sub(rewardTokenAdded));
+    }
+
+    function _getMinimumTokensForRewardSale(uint256 tradeSize) private view returns(uint256 minTokens) {
+        IUniswapV2Pair pair = IUniswapV2Pair(address(uniswapTokenExchange));
+
+        // Get reserves and order them correctly
+        (uint256 reserve0, uint256 reserve1,) = pair.getReserves();
+        (uint256 rewardReserve, uint256 foreignReserve) = 
+            rewardToken == pair.token0() ? (reserve0, reserve1) : (reserve1, reserve0);
+        
+        // Price of our trade if the price didn't slip
+        uint256 immediatePrice = tradeSize.mul(foreignReserve).div(rewardReserve);
+        // The lowest we'd be willing to accept, based on max slippage
+        uint256 acceptablePrice = immediatePrice.mul(10000 - MAX_SLIPPAGE_BASIS_POINTS).div(10000);
+        // Preflight here mostly for the benefit of unit tests
+        require(uniswapRouter.getAmountOut(tradeSize, rewardReserve, foreignReserve) >= acceptablePrice, 
+            "Slippage too high for automagic reinvestment");
```

Comment:
I don't think we're wizards... I think you can remove the term "automagic" altogether.